tests/test_upload.py

```diff
@@ -270,25 +285,69 @@ def _start_procs(*args):
             proc.start()
             yield proc
 
-    def _wait_for_upload(procs, total):
+    def _wait_for_upload(procs, total, optional_files, skipped):
         """Mock the _wait_for_upload function"""
         assert total == num_procs
         for proc in procs:
             proc.stop()
+        return []
 
     monkeypatch.setattr(upload, "_start_procs", _start_procs)
     monkeypatch.setattr(upload, "_wait_for_upload", _wait_for_upload)
-    monkeypatch.setattr(upload, "_compose_file_mapping", lambda *args: [0] * num_procs)
+    monkeypatch.setattr(
+        upload, "_compose_file_mapping", lambda *args: ([0] * num_procs, [])
+    )
 
-    upload._gsutil_assay_upload(None, "")
+    upload._gsutil_assay_upload(
+        api.UploadInfo(
+            JOB_ID,
+            JOB_ETAG,
+            GCS_BUCKET,
+            URL_MAPPING,
+            EXTRA_METADATA,
+            GCS_FILE_MAP,
+            OPTIONAL_FILES,
+            UPLOAD_TOKEN,
+        ),
+        "",
+    )
 
     # Ensure that _gsutil_assay_upload has started and waited for every process
     for proc in process_mocks:
         proc.start.assert_called()
         proc.stop.assert_called()
 
 
-def test_compose_file_mapping(tmpdir):
+def test_compose_file_mapping(tmpdir, monkeypatch):
+    xlsx = str(tmpdir.join("bar.xlsx"))
+
+    failing_map = {str(tmpdir.join("foo.bar")): "foo.bar"}
+    upload_job = api.UploadInfo(
+        JOB_ID,
+        JOB_ETAG,
+        GCS_BUCKET,
+        failing_map,
+        EXTRA_METADATA,
+        GCS_FILE_MAP,
+        OPTIONAL_FILES,
+        UPLOAD_TOKEN,
+    )
+    with pytest.raises(Exception, match="Could not locate file"):
+        upload._compose_file_mapping(upload_job, xlsx)
+
+    # doesn't fail if passed as an optional_file
+    upload_job = api.UploadInfo(
+        JOB_ID,
+        JOB_ETAG,
+        GCS_BUCKET,
+        failing_map,
+        EXTRA_METADATA,
+        GCS_FILE_MAP,
+        OPTIONAL_FILES + list(failing_map.keys()),
+        UPLOAD_TOKEN,
+    )
+    upload._compose_file_mapping(upload_job, xlsx)
```

Comment:
This should also check that the output of `upload._compose_file_mapping` is as expected, e.g. that `foo.bar` is in the skipped files.

Reply:
in next commit